### PR TITLE
Implement canonical linked list cycle checker

### PR DESCRIPTION
**Pull request by caldwecr**
Ok so this definitely looks like I'm new to python b/c I'm using lowest order flow control and logic organization. Show me the light.

### CODE REVIEW

File: challenges/cyclical_linked_list.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    else:
+        return is_cyclical(head, head.next.next)
+
+def is_terminated_soon(head):
+    if head.next is None:
+        return True
+    elif head.next.next is None:
+        return True
+    else:
+        return False
```
Comment (risforrob) [inline challenges/cyclical_linked_list.py:12]:
```python
def is_terminated_soon(head):
    return None in (head, head.next, head.next.next)
```

File: challenges/cyclical_linked_list.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+def has_cycle(head):
+    if is_terminated_soon(head):
+        return False
+    else:
+        return is_cyclical(head, head.next.next)
```
Comment (risforrob) [inline challenges/cyclical_linked_list.py]:
```python
def has_cycle(head):
    return is_terminated_soon(head) and is_cyclical(head, head.next.next)
```

File: challenges/cyclical_linked_list.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    else:
+        return False
+
+def is_cyclical(hop, hophop):
+    if hop == hophop:
+        return True
+    elif is_terminated_soon(hophop):
+        return False
+    else:
+        return is_cyclical(hop.next, hophop.next.next)
```
Comment (risforrob) [inline challenges/cyclical_linked_list.py:26]: `is_cyclical` could be compressed into a one-liner, but I think from a readability standpoint the `if/elif/else` solution is easier to comprehend.

File: challenges/cyclical_linked_list.py (hunk 2 above)
Comment (caldwecr) [reply challenges/cyclical_linked_list.py]: That won't work. is_terminated_soon being True means has_cycle needs to short circuit to false.

File: challenges/cyclical_linked_list.py (hunk 1 above)
Comment (caldwecr) [reply challenges/cyclical_linked_list.py:12]: Sadly this no worko.

```
  1) has cycle when the list contains one node when the node is the end of the list it is False
     Failure/Error: spec/challenges/cyclical_linked_list_spec.py expect(cyclical_linked_list.has_cycle(self.node3)).to(be_false)
         AttributeError: 'NoneType' object has no attribute 'next'
             
     File "./challenges/cyclical_linked_list.py", line 8, in is_terminated_soon
         return None in (head, head.next, head.next.next)
     File "./challenges/cyclical_linked_list.py", line 2, in has_cycle
         if is_terminated_soon(head):
     File "spec/challenges/cyclical_linked_list_spec.py", line 27, in 00000005__it is False
         expect(cyclical_linked_list.has_cycle(self.node3)).to(be_false)
```

Comment (caldwecr) [reply challenges/cyclical_linked_list.py:12]: The expressions in the tuple are all eagerly evaluated, it seems.

Comment (risforrob) [reply challenges/cyclical_linked_list.py:12]: ah, good point, all items are evaluated.

Comment (risforrob) [reply challenges/cyclical_linked_list.py:12]:
```python
def is_terminated_soon(head):
    return head and head.next and head.next.next
```
Which is arguably more concise anyways.

Comment (risforrob) [reply challenges/cyclical_linked_list.py:12]: And should properly delay evaluation.

**Pull request merged**